Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type
+
+-- Lean treats the universal quantifier as an implication.
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+-- An implication is something like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · apply h₀
```

Comment:
I'm probably nitpicking, but it is good practice to use `exact` instead of `apply` when we have a term that matches the goal exactly. If you think this might be confusing to beginners then ignore this.

Reply:
True, MIL for some reason was doing it a lot in these sections. I cleared up some but not all of them, I'll fix it